rts/Sim/Path/QTPFS/PathSearch.cpp

```diff
@@ -146,96 +206,274 @@ bool QTPFS::PathSearch::Execute(unsigned int searchStateOffset) {
+	auto& comp = systemGlobals.GetSystemComponent<PathSpeedModInfoSystemComponent>();
 
 	// be as optimistic as possible: assume the remainder of our path will
 	// cover only flat terrain with maximum speed-modifier between nxtPoint
 	// and tgtPoint
 	switch (searchType) {
-		// This guarantees the best path, but overestimates distance costs considerabily.
+		// This, by default, guarantees the best path, but underestimates distance costs considerabily.
+		// Searching more quads than neccessary is an impact to performance. So a slight increase to the
+		// hcost can be applied to help. Care needs to be taken because if it goes too far, then
+		// performance can drop further instead as the search shifts to best-first-search.
 		case PATH_SEARCH_ASTAR:
-			hCostMult = 1.0f / ( comp.maxRelSpeedMod[nodeLayer->GetNodelayer()] );
+			{
+				const float maxSpeedMod = comp.relSpeedModinfos[nodeLayer->GetNodelayer()].max;
+				const float meanSpeedMod = comp.relSpeedModinfos[nodeLayer->GetNodelayer()].mean;
+				const float chosenSpeedMod = maxSpeedMod - (maxSpeedMod - meanSpeedMod) * modInfo.pfHcostMult;
+				hCostMult = 1.0f / chosenSpeedMod;
+			}
 			break;
 		case PATH_SEARCH_DIJKSTRA:
 			hCostMult = 0.0f;
 			break;
 	}
+}
+
+void QTPFS::PathSearch::RemoveOutdatedOpenNodesFromQueue() {
+	// Remove any out-of-date node entries in the queue.
+	for (int i = 0; i < QTPFS::SEARCH_DIRS; ++i) {
+		DirectionalSearchData& searchData = directionalSearchData[i];
+
+		while (!(*searchData.openNodes).empty()) {
+			SearchQueueNode curOpenNode = (*searchData.openNodes).top();
+			assert(searchThreadData->allSearchedNodes[i].isSet(curOpenNode.nodeIndex));
+			curSearchNode = &searchThreadData->allSearchedNodes[i][curOpenNode.nodeIndex];
+			
+			// Check if this node entity is valid
+			if (curOpenNode.heapPriority <= curSearchNode->GetHeapPriority())
+				break;
+
+			// remove the entry
+			(*searchData.openNodes).pop();
+		}
+	}
+}
 
-	// if (nodeLayer->GetNodelayer() == 2) {
-	// 	LOG("%s: maxRelSpeedMod = %f, hCostMult = %f", __func__
-	// 			, comp.maxRelSpeedMod[nodeLayer->GetNodelayer()], hCostMult);
-	// }
+bool QTPFS::PathSearch::IsNodeActive(const SearchNode& curSearchNode) const {
+	// prevNode catches all cases except the starting node.
+	// Path H-cost will avoid reporting unwalked target node as a false positive, but catch starting node.
+	return (curSearchNode.GetPrevNode() != nullptr)
+		|| (curSearchNode.GetPathCost(NODE_PATH_COST_H) != std::numeric_limits<float>::infinity());
+}
 
+void QTPFS::PathSearch::SetForwardSearchLimit() {
 	auto& fwd = directionalSearchData[SearchThreadData::SEARCH_FORWARD];
+	auto& bwd = directionalSearchData[SearchThreadData::SEARCH_BACKWARD];
 
-	searchThreadData->ResetQueue();
-	ResetState(fwd.srcSearchNode, fwd);
-	UpdateNode(fwd.srcSearchNode, nullptr, 0);
+	constexpr float maxRelativeMapAreaToSearch = (1.f/16.f);
+	constexpr float areaToSearchScale = (1.f/8.f);
+	constexpr float scaleForNodeRevisits = 1.1f;
```

Comment:
any particular reason for 8, 16 and 1.1 or are they just eyeballed? a comment would be nice